### ISSUE TITLE
SQLiteDatabaseLockedException caused by TransferDBHelper.update$aws_storage_s3_release

### ISSUE THREAD

**Issue by yaroslav-v** [aws-amplify/amplify-android#2720]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-android/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-android/discussions).

### Language and Async Model

Java

### Amplify Categories

Storage

### Gradle script dependencies

<details>

```groovy
// Put output below this line
implementation "com.amplifyframework:core:2.14.10"
implementation "com.amplifyframework:aws-storage-s3:2.14.10"
implementation "com.amplifyframework:aws-auth-cognito:2.14.10"

```

</details>

### Environment information

<details>

```
# Put output below this line
------------------------------------------------------------
Gradle 8.2
------------------------------------------------------------

Build time:   2023-06-30 18:02:30 UTC
Revision:     5f4a070a62a31a17438ac998c2b849f4f6892877

Kotlin:       1.8.20
Groovy:       3.0.17
Ant:          Apache Ant(TM) version 1.10.13 compiled on January 4 2023
JVM:          17.0.10 (MacPorts 17.0.10+11)
OS:           Mac OS X 14.3 x86_64

```

</details>

### Describe the bug

Hi there!

We have a few reports (about 20 at the moment) on Firebase with this crash for a few latest releases of the library v2.14.x including the last one v2.14.10. The issue appears on Android 10-13, affected devices are mostly Motorola (around 80%) with some Samsung, Xiaomi etc.

The issue is quite rare and, unfortunately, there are no exact steps to reproduce it. I don't see any specific pattern in accompanying logs. Sometimes it happened after the app update, but in the minority of the cases.

### Code Snippet

```java
// Initialization in Application class

Amplify.Logging.disable();

Amplify.addPlugin(new AWSCognitoAuthPlugin());
Amplify.addPlugin(new AWSS3StoragePlugin());

AmplifyConfiguration amplifyConfig = AmplifyConfiguration.builder(context)
		.devMenuEnabled(false)
		.build();
Amplify.configure(amplifyConfig, context);


// Uploading, somewhere in the app

DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ", Locale.US);
dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));

Map<String, String> metadata = new ArrayMap<>(4);
metadata.put("date", dateFormat.format(date));
metadata.put("session_uuid", uid);
metadata.put("token", authToken);
metadata.put("service", serviceName);

StorageUploadFileOptions options =
		StorageUploadFileOptions.builder()
				.accessLevel(StorageAccessLevel.PUBLIC)
				.contentType(mimeType)
				.metadata(metadata)
				.build();

final File file = new File(path);
if (file.canRead()) {
	Amplify.Storage.uploadFile(
			file.getName(),
			file,
			options,
			result -> {
				// write success logs
			},
			failure -> {
				// write failure logs
			}
	);
}

```

### Log output

<details>

```
// Put your logs below this line
Fatal Exception: android.database.sqlite.SQLiteDatabaseLockedException
database is locked (code 5 SQLITE_BUSY)
android.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount (SQLiteConnection.java)
android.database.sqlite.SQLiteConnection.executeForChangedRowCount (SQLiteConnection.java:900)
android.database.sqlite.SQLiteSession.executeForChangedRowCount (SQLiteSession.java:756)
android.database.sqlite.SQLiteStatement.executeUpdateDelete (SQLiteStatement.java:66)
android.database.sqlite.SQLiteDatabase.updateWithOnConflict (SQLiteDatabase.java:1817)
android.database.sqlite.SQLiteDatabase.update (SQLiteDatabase.java:1763)
com.amplifyframework.storage.s3.transfer.TransferDBHelper.update$aws_storage_s3_release (TransferDBHelper.kt:192)
com.amplifyframework.storage.s3.transfer.TransferDB.updateState (TransferDB.kt:215)
com.amplifyframework.storage.s3.transfer.TransferStatusUpdater.updateTransferState (TransferStatusUpdater.kt:106)
com.amplifyframework.storage.s3.transfer.TransferWorkerObserver.updateTransferState (TransferWorkerObserver.kt:166)
com.amplifyframework.storage.s3.transfer.TransferWorkerObserver.handleTransferStatusUpdate (TransferWorkerObserver.kt:116)
com.amplifyframework.storage.s3.transfer.TransferWorkerObserver.access$handleTransferStatusUpdate (TransferWorkerObserver.kt:35)
com.amplifyframework.storage.s3.transfer.TransferWorkerObserver$onChanged$1.invokeSuspend (TransferWorkerObserver.kt:95)
kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (ContinuationImpl.kt:33)
kotlinx.coroutines.DispatchedTask.run (DispatchedTask.kt:108)
java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
java.lang.Thread.run (Thread.java:923)

```

</details>

### GraphQL Schema

<details>

```graphql
// Put your schema below this line


```

</details>

**Comment by ankpshah** [comment]
Hello thanks for reporting this issue. Someone from our team will take look at this.

**Comment by mattcreaser** [comment]
🤔 This error suggests that there may be concurrent SqQliteOpenHelper instances access the database. We will need to look into how that might be possible.

**Comment by yaroslav-v** [comment]
Hi guys!

I just want to give you some additional context. For the last 30 days this issue was registered for 9 users in Firebase.

The issue appears on Android 7-13, affected devices are Samsung (S10, A13, 
Galaxy A72), Sony (Xperia 8), Motorola (Moto G Pure), Xiaomi (Redmi 5), Oppo, LG.

The stack trace is the same for all cases, except for some cases the cause is "database is locked (code 5 SQLITE_BUSY)" and for others it's "database is locked (code 5 SQLITE_BUSY[5])".

AWS Amplify v2.14.11

**Comment by mattcreaser** [comment]
This fix has been released in Amplify **2.15.2** 🎉 

Closing this issue, please feel free to open a new one if required.
